internal/sandbox/fs.go

```diff
@@ -1,511 +1,547 @@
+	}
+	f, err := Open(fsys.base, name)
+	if err != nil {
+		return nil, fsError("open", name, err)
 	}
-	return stat, errno
+	return &fsFile{fsys: fsys.base, File: f}, nil
 }
 
-func (f *throttleFile) FDFileStatSetSize(ctx context.Context, size wasi.FileSize) wasi.Errno {
-	errno := f.base.FDFileStatSetSize(ctx, size)
-	if canThrottle(errno) {
-		f.fsys.throttleWrite(ctx, int64(size))
+func (fsys *fsFileSystem) Stat(name string) (fs.FileInfo, error) {
+	if !fs.ValidPath(name) {
+		return nil, fsError("stat", name, fs.ErrNotExist)
 	}
-	return errno
+	s, err := Stat(fsys.base, name)
+	if err != nil {
+		return nil, fsError("stat", name, err)
+	}
+	return &fsFileInfo{name: path.Base(name), stat: s}, nil
 }
 
-func (f *throttleFile) FDFileStatSetTimes(ctx context.Context, accessTime, modifyTime wasi.Timestamp, flags wasi.FSTFlags) wasi.Errno {
-	errno := f.base.FDFileStatSetTimes(ctx, accessTime, modifyTime, flags)
-	if canThrottle(errno) {
-		f.fsys.throttleWrite(ctx, 1)
+func fsError(op, name string, err error) error {
+	err = unwrap(err)
+	switch {
+	case errors.Is(err, EEXIST):
+		err = fs.ErrExist
+	case errors.Is(err, ENOENT):
+		err = fs.ErrNotExist
+	case errors.Is(err, EINVAL):
+		err = fs.ErrInvalid
+	case errors.Is(err, EBADF):
+		err = fs.ErrClosed
 	}
-	return errno
+	return &fs.PathError{Op: op, Path: name, Err: err}
 }
 
-func (f *throttleFile) FDPread(ctx context.Context, iovecs []wasi.IOVec, offset wasi.FileSize) (wasi.Size, wasi.Errno) {
-	n, errno := f.base.FDPread(ctx, iovecs, offset)
-	if canThrottle(errno) {
-		f.fsys.throttleRead(ctx, sizeToInt64(n))
+func unwrap(err error) error {
+	if e := errors.Unwrap(err); e != nil {
+		err = e
 	}
-	return n, errno
+	return err
 }
 
-func (f *throttleFile) FDPwrite(ctx context.Context, iovecs []wasi.IOVec, offset wasi.FileSize) (wasi.Size, wasi.Errno) {
-	n, errno := f.base.FDPwrite(ctx, iovecs, offset)
-	if canThrottle(errno) {
-		f.fsys.throttleWrite(ctx, sizeToInt64(n))
-	}
-	return n, errno
+var (
+	_ fs.StatFS = (*fsFileSystem)(nil)
+)
+
+type fsFile struct {
+	fsys FileSystem
+	dir  *dirbuf
+	File
 }
 
-func (f *throttleFile) FDRead(ctx context.Context, iovecs []wasi.IOVec) (wasi.Size, wasi.Errno) {
-	n, errno := f.base.FDRead(ctx, iovecs)
-	if canThrottle(errno) {
-		f.fsys.throttleRead(ctx, sizeToInt64(n))
+func (f *fsFile) Stat() (fs.FileInfo, error) {
+	stat, err := f.File.Stat("", AT_SYMLINK_NOFOLLOW)
+	if err != nil {
+		return nil, err
 	}
-	return n, errno
+	name := path.Base(f.File.Name())
+	return &fsFileInfo{name: name, stat: stat}, nil
 }
 
-func (f *throttleFile) FDWrite(ctx context.Context, iovecs []wasi.IOVec) (wasi.Size, wasi.Errno) {
-	n, errno := f.base.FDWrite(ctx, iovecs)
-	if canThrottle(errno) {
-		f.fsys.throttleWrite(ctx, sizeToInt64(n))
+func (f *fsFile) Read(b []byte) (int, error) {
+	iovs := [][]byte{b}
+	read := 0
+	for {
+		n, err := f.File.Readv(iovs)
+		if n > 0 {
+			read += n
+		}
+		if read == len(b) {
+			return read, nil
+		}
```

Comment:
Is there a chance here and elsewhere that an error might be suppressed because `read == len(b)`?

Reply:
Yes, I should move this condition after the error check 👍